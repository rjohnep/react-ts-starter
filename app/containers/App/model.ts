import { UserT } from './types';

export const initialValue: UserT = {
  firstName: 'Jan',
  lastName: 'Kowalski',
  roles: ['admin']
};

export const setName = (state: UserT, firstName: string): UserT => ({
  ...state,
  firstName
});


export const mergeUsers = (currentUser: UserT, newUser: UserT): UserT => ({
  ...currentUser,
  ...newUser,
  ...((currentUser.roles || newUser.roles) && {
    roles: [
      ...(currentUser.roles || []),
      ...(newUser.roles || [])
    ]
  })
});

// *** 'a few time ago' area *** //

// const creatorProfileProperties = {
//   subscription: 'Medium',
//   roles: ['editor', 'subscriber']
// };

// export const mergeUsers = (currentUser: UserT, newUser: UserT): UserT => ({
//   ...currentUser,
//   ...newUser,
//   ...((currentUser.roles || newUser.roles) && {
//     roles: {
//       // ERROR: object instead of array
//       ...(currentUser.roles || []),
//       ...(newUser.roles || [])
//     }
//   })
//   // ...creatorProfileProperties
// });
