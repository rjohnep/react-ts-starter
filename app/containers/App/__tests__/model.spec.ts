import { initialValue, setName, mergeUsers } from '../model';

const userMock = initialValue;

describe('App model', () => {
  it('should update a name of an user', () => {
    const fixture = 'Jakub';

    expect(setName(userMock, fixture)).toHaveProperty('firstName', fixture);
  });

  it('should merge users correctly, with deep merging of roles', () => {
    const newUser = {
      ...userMock,
      roles: ['user']
    };

    // main pitfall
    expect(mergeUsers(userMock, newUser)).toHaveProperty(
      'roles',
      [...(userMock.roles || []), ...newUser.roles]
    );
    // simplify testing all properties
    expect(mergeUsers(userMock, newUser)).toMatchSnapshot();
  });
});
