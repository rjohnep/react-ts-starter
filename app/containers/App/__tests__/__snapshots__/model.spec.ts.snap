// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App model should merge users correctly, with deep merging of roles 1`] = `
Object {
  "firstName": "Jan",
  "lastName": "Kowalski",
  "roles": Array [
    "admin",
    "user",
  ],
}
`;
